### PR TITLE
bears/go: Add GoErrCheckBear

### PR DESCRIPTION
**Pull request by mr-karan**
Used errcheck to find function calls
that have unchecked errors in Go.

Fixes https://github.com/coala-analyzer/coala-bears/issues/31

### CODE REVIEW

**Comment by sils1297** [comment]
ack https://github.com/coala-analyzer/coala-bears/pull/443/commits/1fb4db6376436515e4cc1baa1e02c9b2be688d35 I like it, assuming there's no way to configure this further?

**Comment by mr-karan** [comment]
https://asciinema.org/a/45798

**Comment by sils1297** [comment]
the message is nonexplanatory, you're already working on that, right? Looks good otherwise.

**Comment by SanketDG** [comment]
TODO upgrade coala in requirements.txt

**Comment by mr-karan** [comment]
Should I wait for https://github.com/coala-analyzer/coala-bears/pull/446 to be merged ?

File: bears/go/GoErrCheckBear.py [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+        result_message="Erroneous function calls detected by errcheck")
+class GoErrCheckBear:
+    """
+    Checks the code for all function calls that have unchecked errors.
+    GoErrCheckBear runs ``errcheck`` over each file to find such functions.
+    """
+    LANGUAGES = "Go"
+
+    @staticmethod
+    def create_arguments(filename, file, config_file):
```
Comment (arafsheikh) [inline bears/go/GoErrCheckBear.py]: `errcheck` takes in a few arguments. Implementing them as settings will make this bear more complete.

**Comment by SanketDG** [comment]
@mr-karan this looks good.

And if you want, you can implement more settings (which would be awesome) or go ahead and open an issue for that.

**Comment by mr-karan** [comment]
@SanketDG I would have gladly worked on it, but as i told on my Ubuntu system it randomly stopped working, I am unable to figure out why. If someone can help me with new asciinema and/or the settings (which is just a ` -blank` flag), would be great.

**Comment by sils1297** [comment]
It randomly stopped working? Do you have a second system around to test maybe? Latest version installed?

File: bears/go/GoErrCheckBear.py [hunk 2]
```diff
@@ -0,0 +1,18 @@
+from coalib.bearlib.abstractions.Linter import linter
+
+
+@linter(executable='errcheck',
+        output_format='regex',
+        output_regex=r'(?P<filename>[^:]+):(?P<line>\d+):'
+        r'(?P<col>\d+)\s*(?P<message>.*)',
+        result_message="Erroneous function calls detected by errcheck")
```
Comment (sils1297) [inline bears/go/GoErrCheckBear.py]: I don't htink the message is particularly good, maybe rather something like "This function call has an unchecked error."?

**Comment by sils1297** [comment]
unack https://github.com/coala-analyzer/coala-bears/pull/443/commits/fb347bc21661c25ddd7d680dc18c011f7680d32e

File: bears/go/GoErrCheckBear.py (hunk 2 above)
Comment (mr-karan) [inline bears/go/GoErrCheckBear.py]: I am planning to install Arch tonight on my system, if you can wait till tomorrow I can do it :+1:

**Comment by Adrianzatreanu** [comment]
marking as WIP as it requires a rebase and it is incomplete, it should get a few more arguments.

**Comment by Adrianzatreanu** [comment]
unack 5971382

File: requirements.txt [hunk 3]
```diff
@@ -27,3 +27,4 @@ vim-vint~=0.3.6
 # Use >=3.3 once nltk-3.3 is released.
 nltk~=3.1.0
 appdirs~=1.4.0
+
```
Comment (Adrianzatreanu) [inline requirements.txt]: it didnt have a newline at end of file before? or why is this needed in this commit?

Comment (mr-karan) [inline requirements.txt]: rah, i don't know where did this creep in. Fixed. This is needed because @Makman2 did some changes which are required for tests to pass here. The `result_message` was added

**Comment by Adrianzatreanu** [comment]
ack dc2d7d5

**Comment by mr-karan** [comment]
https://asciinema.org/a/46366

**Comment by Adrianzatreanu** [comment]
ack 443e064

**Comment by mr-karan** [comment]
reack https://github.com/coala-analyzer/coala-bears/pull/443/commits/ac142871ad5009ca510e015573538d08c3541ada

**Comment by SanketDG** [comment]
TODO upgrade coala in requirements.txt

**Comment by SanketDG** [comment]
@mr-karan needs rebase. Also the coala upgrade should be in a different commit.

File: tests/go/GoErrCheckBearTest.py [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+f.Close()
+}
+""".splitlines(keepends=True)
+
+GoErrCheckBearTest = verify_local_bear(GoErrCheckBear,
+                                       valid_files=(good_file,),
+                                       invalid_files=(bad_file,),
+                                       tempfile_kwargs={"suffix": ".go"})
+
+GoErrCheckBearWithSetting = verify_local_bear(GoErrCheckBear,
```
Comment (AbdealiJK) [inline tests/go/GoErrCheckBearTest.py]: These class names don't end with "Test"
Last I remember, pytest finds classes ending with "Test"
Are you sure these tests are even being run ?

Comment (AbdealiJK) [inline tests/go/GoErrCheckBearTest.py]: Also, don't use the same name for all test classes ... get's difficult to debug

Comment (mr-karan) [inline tests/go/GoErrCheckBearTest.py]: Noted. Thanks, this was the issue indeed.

**Comment by mr-karan** [comment]
reack https://github.com/coala-analyzer/coala-bears/pull/443/commits/d5ad1584d0af7c2e08614b8a26b292e9e2a97487

**Comment by sils1297** [comment]
unack d5ad158 as per discussion

File: bears/go/GoErrCheckBear.py [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+
+    @staticmethod
+    def create_arguments(filename, file, config_file,
+                         ignore: str='',
+                         ignorepkg: str='',
+                         ignoretests: str='',
+                         tags: str='',
+                         asserts: str='',
+                         blank: str=''
+                         ):
```
Comment (Makman2) [inline bears/go/GoErrCheckBear.py]: please put the `):` into the line before^^

File: bears/go/GoErrCheckBear.py [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+        :param ignorepkg:    Takes a comma-separated list of package
+                             import paths to ignore.
+        :param asserts:      Enables checking for ignored type assertion results
+        :param blank:        Enables checking for assignments of errors to
+                             the blank identifier.
+
+        """
+        args = ()
+        if ignore:
+            args += ("-ignore", "'" + (ignore) + "'",)
```
Comment (Makman2) [inline bears/go/GoErrCheckBear.py]: redundant comma at end

File: bears/go/GoErrCheckBear.py [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+        if ignorepkg:
+            args += ("-ignorepkg", "'" + (ignorepkg) + "'",)
```
Comment (Makman2) [inline bears/go/GoErrCheckBear.py]: also redundant comma at end

File: requirements.txt [hunk 8]
```diff
@@ -1,5 +1,5 @@
 # Use >= for development versions so that source builds always work
-coala>=0.7.0.dev20160518085900
+coala>=0.7.0.dev20160522150811
```
Comment (Makman2) [inline requirements.txt]: why do you upgrade? was there an API change?

Comment (mr-karan) [inline requirements.txt]: @Makman2 I believe result message was added in the Linter API, so I upgraded

File: bears/go/GoErrCheckBear.py [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+        r'(?P<col>\d+)\s*(?P<message>.*)',
+        result_message="This function call has an unchecked error.")
+class GoErrCheckBear:
+
+    """
+    Checks the code for all function calls that have unchecked errors.
+    GoErrCheckBear runs ``errcheck`` over each file to find such functions.
+
+    For more information on the analysis visit <https://github.com/kisielk/errcheck>
+
```
Comment (sils1297) [inline bears/go/GoErrCheckBear.py]: unneeded newline

File: bears/go/GoErrCheckBear.py [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+    @staticmethod
+    def create_arguments(filename, file, config_file,
+                         ignore: str='',
+                         ignorepkg: str='',
+                         ignoretests: str='',
+                         tags: str='',
+                         asserts: str='',
+                         blank: str=''):
+        """
+        Bear configuration arguments.
```
Comment (sils1297) [inline bears/go/GoErrCheckBear.py]: @Makman2 how does this behave on doc merge?

File: bears/go/GoErrCheckBear.py [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+        args = ()
+        if ignore:
+            args += ("-ignore", "'" + (ignore) + "'")
+        if ignorepkg:
+            args += ("-ignorepkg", "'" + (ignorepkg) + "'")
+        if blank:
+            args += ("-blank",)
+        if asserts:
+            args += ("-asserts",)
+        print(args)
```
Comment (sils1297) [inline bears/go/GoErrCheckBear.py]: debug print

File: bears/go/GoErrCheckBear.py [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+        """
+        args = ()
+        if ignore:
+            args += ("-ignore", "'" + (ignore) + "'")
+        if ignorepkg:
+            args += ("-ignorepkg", "'" + (ignorepkg) + "'")
+        if blank:
+            args += ("-blank",)
+        if asserts:
+            args += ("-asserts",)
```
Comment (sils1297) [inline bears/go/GoErrCheckBear.py]: maybe you could rather make a dict giving each of those args it's value and then you just `args += sum(key for key, value in arg_dict if value`

File: tests/go/GoErrCheckBearTest.py [hunk 13]
```diff
@@ -0,0 +1,107 @@
...
+)
+
+func main() {
+  f, _ := os.Open("foo")
+  log.Println("opened file")
+  io.Copy(os.Stdout, f)
+}""".splitlines(keepends=True)
+
+GoErrCheckBearTest = verify_local_bear(GoErrCheckBear,
+                                       valid_files=(good_file,),
```
Comment (sils1297) [inline tests/go/GoErrCheckBearTest.py]: can you add the other two files to valid files here?

File: requirements.txt (hunk 8 above)
Comment (Makman2) [inline requirements.txt]: a rebase should suffice, the `result_message` update was already merged :+1:

File: bears/go/GoErrCheckBear.py [hunk 14]
```diff
@@ -0,0 +1,55 @@
+
+from coalib.bearlib.abstractions.Linter import linter
+from coalib.results.RESULT_SEVERITY import RESULT_SEVERITY
```
Comment (Adrianzatreanu) [inline bears/go/GoErrCheckBear.py]: where are you using this? isnt it an useless import?

File: bears/go/GoErrCheckBear.py [hunk 15]
```diff
@@ -0,0 +1,55 @@
...
+    LANGUAGES = "Go"
+
+    @staticmethod
+    def create_arguments(filename, file, config_file,
+                         ignore: str='',
+                         ignorepkg: str='',
+                         ignoretests: str='',
+                         tags: str='',
+                         asserts: str='',
+                         blank: str=''):
```
Comment (Adrianzatreanu) [inline bears/go/GoErrCheckBear.py]: not all arguments are within ``:param x:``

File: requirements.txt (hunk 8 above)
Comment (Adrianzatreanu) [inline requirements.txt]: this does not fit in this commit anyway. if its a must and you need to do it anyway, please do it in a different commit

File: bears/go/GoErrCheckBear.py [hunk 16]
```diff
@@ -0,0 +1,50 @@
+from coalib.bearlib.abstractions.Linter import linter
+from coalib.results.Result import Result
+from coalib.settings.Setting import path
+
+
+@linter(executable='errcheck',
+        output_format='regex',
+        output_regex=r'(?P<filename>[^:]+):(?P<line>\d+):'
+        r'(?P<col>\d+)\s*(?P<message>.*)',
```
Comment (Adrianzatreanu) [inline bears/go/GoErrCheckBear.py]: this is probably not a mistake, but from what i've seen on other bears, the ``r`` should be under the ``r`` on the previous line

File: bears/go/GoErrCheckBear.py [hunk 17]
```diff
@@ -0,0 +1,50 @@
...
+        result_message="This function call has an unchecked error.")
+class GoErrCheckBear:
+    """
+    Checks the code for all function calls that have unchecked errors.
+    GoErrCheckBear runs ``errcheck`` over each file to find such functions.
+
+    For more information on the analysis visit <https://github.com/kisielk/errcheck>
```
Comment (Adrianzatreanu) [inline bears/go/GoErrCheckBear.py]: add a period please

**Comment by Adrianzatreanu** [comment]
some small things and imo its good to go! :)

File: tests/go/GoErrCheckBearTest.py [hunk 18]
```diff
@@ -0,0 +1,109 @@
...
+func main() {
+  f, _ := os.Open("foo")
+  log.Println("opened file")
+  io.Copy(os.Stdout, f)
+}""".splitlines(keepends=True)
+
+GoErrCheckBearTest = verify_local_bear(GoErrCheckBear,
+                                       valid_files=(
+                                           good_file, assert_file,
+                                           blank_file,),
```
Comment (sils1297) [inline tests/go/GoErrCheckBearTest.py]: what's with the ignorepkg_file here?

File: tests/go/GoErrCheckBearTest.py [hunk 19]
```diff
@@ -0,0 +1,109 @@
...
+                                       invalid_files=(bad_file,),
+                                       tempfile_kwargs={"suffix": ".go"})
+
+GoErrCheckBearWithIgnoreTest = verify_local_bear(GoErrCheckBear,
+                                                 valid_files=(good_file,),
+                                                 invalid_files=(bad_file,),
```
Comment (sils1297) [inline tests/go/GoErrCheckBearTest.py]: shouldn't one of the files tested in the previous test go invalid? But none of them are listed here

File: tests/go/GoErrCheckBearTest.py [hunk 20]
```diff
@@ -0,0 +1,109 @@
...
+                                                        ignorepkg_file,),
+                                                    settings={
+                                                        "ignorepkg":
+                                                        'io'},
+                                                    tempfile_kwargs={"suffix":
+                                                                     ".go"})
+
+GoErrCheckBearWithBlankTest = verify_local_bear(GoErrCheckBear,
+                                                valid_files=(good_file,),
+                                                invalid_files=(bad_file,),
```
Comment (sils1297) [inline tests/go/GoErrCheckBearTest.py]: don't you wanna put the blank file here?

File: bears/go/GoErrCheckBear.py [hunk 21]
```diff
@@ -0,0 +1,49 @@
...
+
+        """
+        args = ()
+        if ignore:
+            args += ("-ignore", "'" + (ignore) + "'")
+        if ignorepkg:
+            args += ("-ignorepkg", "'" + (ignorepkg) + "'")
+        if blank:
+            args += ("-blank",)
+        if asserts:
```
Comment (sils1297) [inline bears/go/GoErrCheckBear.py]: hey this is a bool, right? Why is it a str and not a bool in the paramete list? Then you can also give `true` `false` `yeah` `nope` and so on.

**Comment by sils1297** [comment]
minor changes needed on https://github.com/coala-analyzer/coala-bears/pull/443/commits/263a1f687d741d4279f7423300aa793defe9c204 but we're eheading the right way, also rebase needed. Needs work

**Comment by mr-karan** [comment]
hey so some clarification with the `valid files`

A valid file would mean `errcheck` doesn't report any error. But if I add the `ignorepkg_file` it will report the errors returned by the unignored package and hence not a valid file anymore. Hence I decided to not add in the valid_files tuple for the first case. Similar is the case for blank_files which isn't a valid file as errors checked apart from blank flag are still there. So IMHO it's best to leave it like this as we are testing each file with it's setting in any case. 

Apart from this confusion, I have done changes whatever you told. :+1: 
@sils1297

**Comment by sils1297** [comment]
tests need work and this needs a rebase

File: bears/go/GoErrCheckBear.py [hunk 22]
```diff
@@ -0,0 +1,49 @@
...
+    GoErrCheckBear runs ``errcheck`` over each file to find such functions.
+
+    For more information on the analysis visit <https://github.com/kisielk/errcheck>.
+    """
+    LANGUAGES = "Go"
+
+    @staticmethod
+    def create_arguments(filename, file, config_file,
+                         ignore: str='',
+                         ignorepkg: str='',
```
Comment (sils1297) [inline bears/go/GoErrCheckBear.py]: those are lists, huh? Please use `typed_list(str)` and concatenate them again so coala does the list verification and parsing.

Comment (mr-karan) [inline bears/go/GoErrCheckBear.py]: done @sils1297

**Comment by sils1297** [comment]
ack https://github.com/coala-analyzer/coala-bears/pull/443/commits/77279d3ca7fff4728c2b75f5e203c20b1e9836dd is acceptable though the test could still be improved a bit. I think it's ok though to go in production.

**Comment by sils1297** [comment]
@rultor merge
